File: Data Structures & Algorithms/validate-parentheses/submission-3.cpp
class Solution {
public:
    bool isValid(string s) {
        stack<char> st;
        for(const char& c: s) {
            if(c == '(' or c == '{' or c == '[') {
                st.push(c);
            }
            else {
                if(c == ')') {
                    if(st.size() > 0 and st.top() == '(') {
                        st.pop();
                    }
                    else {
                        return false;
                    }
                }
                if(c == '}') {
                    if(st.size() > 0 and st.top() == '{') {
                        st.pop();
                    }
                    else {
                        return false;
                    }
                }
                if(c == ']') {
                    if(st.size() > 0 and st.top() == '[') {
                        st.pop();
                    }
                    else {
                        return false;
                    }
                }
            }
        }

        return st.size() == 0;
    }
};

File: Data Structures & Algorithms/combination-target-sum-ii/submission-3.py
class Solution:
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        res = []
        candidates.sort()

        def dfs(temp, idx, total):
            if total == target:
                res.append(temp[:])
                return

            if total > target or idx >= len(candidates):
                return 

            for j in range(idx, len(candidates)):
                if j > idx and candidates[j] == candidates[j-1] :
                    continue

                temp.append(candidates[j])
                dfs(temp, j + 1, total + candidates[j])
                temp.pop()

        dfs([], 0, 0)
        return res